File: src/main/java/com/ptah/common/impl/RedisService.java
package com.ptah.common.impl;

import com.ptah.common.CacheService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class RedisService implements CacheService {
    @Autowired
    private RedisTemplate<Object, Object> redisTemplate;

    @Override
    public String getByKey(String key) {
        return getByKey(key, String.class);
    }

    @Override
    public <T> T getByKey(String key, Class<T> type) {
        Object value = redisTemplate.opsForValue().get(key);
        return Optional.ofNullable(value).map(type::cast).orElseThrow();
    }

    @Override
    public void setValue(String key, Object value) {
        redisTemplate.opsForValue().set(key, value);
    }

    @Override
    public void setValue(String key, Object value, long timeout) {
        redisTemplate.opsForValue().set(key, value, timeout);
    }

    @Override
    public boolean hasKey(String key) {
        return redisTemplate.hasKey(key);
    }
}

File: backend/src/main/kotlin/com/ptah/common/impl/RedisService.kt
package com.ptah.common.impl

import com.ptah.common.CacheService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Service
import java.util.*
import java.util.concurrent.TimeUnit

@Service
class RedisService : CacheService {
    @Autowired
    lateinit var redisTemplate: RedisTemplate<Any, Any>
    override fun getByKey(key: String?): String = getByKey(key, String::class.java)

    override fun <T> getByKey(key: String?, type: Class<T>): T {
        val value = redisTemplate.opsForValue()[key!!]
        return Optional.ofNullable(value).map { obj: Any? -> type.cast(obj) }.orElseThrow()
    }

    override fun setValue(key: String, value: Any) {
        redisTemplate.opsForValue()[key] = value
    }

    override fun setValue(key: String, value: Any, timeoutInSeconds: Long) {
        redisTemplate.opsForValue().set(key, value, timeoutInSeconds, TimeUnit.SECONDS)
    }

    override fun hasKey(key: String): Boolean {
        return redisTemplate.hasKey(key)
    }
}
